posts/2017-05-02-io-monad-for-cats.md

```diff
@@ -0,0 +1,404 @@
+
+This is kind of a clever trick.  It allows Haskell to simultaneously be pure *and* still have excellent support for manipulating effects and interacting with the "real world".  But why is it relevant to Scala?  After all, Scala is an impure language.  We don't *need* to go through this complex rigmarole of describing our effects and composing those descriptions; the language lets us *just do it!*  So why wouldn't we just, you know, evaluate the effects that we need evaluated?
+
+The answer is that we want to reason about *where* and *when* our effects are evaluated.  And of course, we want to be able to leverage laws and abstractions which assume equational semantics for expressions (i.e. referential transparency).  Cats is full of these sorts of abstractions, and cats-laws provides a vast set of laws which describe them.  But all of these abstractions and all of these laws break down the *moment* you introduce some sort of side-effecting expression.  Because, much like our referential transparency example from earlier, these abstractions *assume* that you can substitute expressions with their evaluated results, and that's just not true in the presence of side-effects.
+
+What we need is a data type which allows us to encapsulate Scala-style side-effects in the form of a *pure* value, on which referential transparency holds and which we can compose using other well-defined abstractions, such as `Monad`.  Scalaz defines two such data types which meet these criteria: `scalaz.effect.IO` and `scalaz.concurrent.Task`.  But in practice, nearly everyone uses `Task` instead of `IO` because of its support for *asynchronous* effects.
+
+Cats does not define any such abstraction, and what's worse is the cats *ecosystem* also doesn't really provide any such abstraction.  There are two `Task` implementations that are relatively commonly used with cats – namely, `monix.Task` and `fs2.Task` – but these are not part of cats per se, nor are they deeply integrated into its abstraction hierarchy.  Additionally, the proliferation of broadly equivalent options has led to confusion in the ecosystem, as well as some justifiable fear, uncertainty and doubt on the part of those evaluating cats as a potential dependency.
```

Comment:
Perhaps the phrasing could be softened though. As you said, it does have the potential to fan some flames.